hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/AbstractFSWAL.java

```diff
@@ -1190,9 +1201,18 @@ public OptionalLong getLogFileSizeIfBeingWritten(Path path) {
     try {
       Path currentPath = getOldPath();
       if (path.equals(currentPath)) {
+        // Currently active path.
         W writer = this.writer;
         return writer != null ? OptionalLong.of(writer.getSyncedLength()) : OptionalLong.empty();
       } else {
+        W temp = inflightWALClosures.get(path.getName());
+        if (temp != null) {
```

Comment:
We wont be having the rollWriterLock based lock in WAL close area.  It might be possible that the check pass here and by when it do temp.getSyncedLength(), the WAL might be closed.  That will be ok?  Good to double check.

Reply:
Not sure what you mean, we are returning a long from getSyncedLength() under the same lock (we are not returning the reference temp), actual method getSyncedLength() is thread-safe, if thats what you are asking..

Also, both shutdown() / rollWriter are under the scope of rollWriterLock.

Reply:
Sorry for not making clear.
What we are in effect doing is, when the WAL file is not yet closed, we get the length from writer which tracks the synced length.  Once the WAL file is closed, (we know that from the output of this method) there is different logic.   (I have not seen that full area though).  Now with the async close, it is possible that when this method execution is at Line 1208, the close is not over but we return from here, the WAL file is really closed.   Still we will return a value and system thinks WAL file is still active one and undergo related logic.   It may be very much fine in those places. But I just wanted to make we double check.
With sync close,  this method and the writer close all were having a common lock. (rollWriterLock).  The entire  rollWriter(boolean force) is under that lock and cur writer close was within it.
Hope am making it clear now? 
Again it may be very much fine only. But just can see.

Reply:
Thanks for clarifying.

> What we are in effect doing is, when the WAL file is not yet closed, we get the length from writer which tracks the synced length. Once the WAL file is closed, 

Correct.

>  Now with the async close, it is possible that when this method execution is at Line 1208, the close is not over but before we return from this method, the WAL file is really closed

Correct

> Still we will return a value and system thinks WAL file is still active one and undergo related logic. It may be very much fine in those places. But I just wanted to make sure we double check.

Yes this is transparently handled by WalEntryStream. If what you said happens (which is possible), WALEntryStream thinks that the file is not fully done yet (in that loop iteration) and in the next attempt when it does the same check again to read more bytes, it sees that the file is already done and closed and either reads or terminates accordingly.

Relevant code links

https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/WALEntryStream.java#L283 (the entry read is null, so it just loops again)

https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/WALEntryStream.java#L197 (this is for looping again, in the subsequent next() call, this fill will be seen as closed and it enters the if block in L180)

Reply:
Thanks for checking and confirming.  Looks good then.

Reply:
Thanks for taking a look @anoopsjohn